File: Backend/routers/message.py
from fastapi import APIRouter
from fastapi.responses import JSONResponse
import requests
from pydantic import BaseModel
from myopenai import *

router = APIRouter()

class StartMessageSchema(BaseModel):
    username: str
    message: str

class MessageSchema(BaseModel):
    username: str
    message: str
    thread_id: str
    
@router.post("/start")
def start_message(message: StartMessageSchema):
    #create thread and send message to GPT-3.5 then run 
    #save chat data to database
    #return thread_id and reply message
    try:
        userInput = message.message
        topic = get_Topic(userInput)
        thread, run = create_thread_and_run(userInput)
        createChat = requests.post(os.getenv("MONGODB") + "/chat/create/" + topic, json={"username": message.username, "content": userInput, "isNotGPT": True, "type": "text", "chatid": thread.id})
        if createChat.status_code != 200:
            return JSONResponse(status_code=createChat.status_code, content="Created Chat Error")
        run = wait_on_run(run, thread)
        GPTresponse = get_GPT_response(thread)
        
        sendResponse = requests.post(os.getenv("MONGODB") + "/message/create/", json={"username": "KU-Assistant", "content": GPTresponse, "isNotGPT": False, "type": "text", "chatid": thread.id})
        if sendResponse.status_code != 200:
            # Should delete created chat above
            return JSONResponse(status_code=sendResponse.status_code, content="Send Response Error")
        return JSONResponse(status_code=200, content={"thread_id": thread.id, "message": GPTresponse})
    except:
        return JSONResponse(status_code=500, content={"error": "Internal Server Error"})

@router.post("/message")
def send_message(message: MessageSchema):
    #send message to GPT-3.5 then run
    #save chat data to database
    #return reply message
    try:
        userInput = message.message
        thread = get_thread(message.thread_id)
        sendUserMessage = requests.post(os.getenv("MONGODB") + "/message/create/", json={"username": message.username, "content": userInput, "isNotGPT": True, "type": "text", "chatid": thread.id})
        if sendUserMessage.status_code != 200:
            return JSONResponse(status_code=sendUserMessage.status_code, content="Send User Message Error")
        run = submit_message(os.getenv("ASSISTANT_ID"), thread, userInput)
        run = wait_on_run(run, thread)
        GPTresponse = get_GPT_response(thread)
        sendResponse = requests.post(os.getenv("MONGODB") + "/message/create/", json={"username": "KU-Assistant", "content": GPTresponse, "isNotGPT": False, "type": "text", "chatid": thread.id})
        if sendResponse.status_code != 200:
            # Should delete created chat above
            return JSONResponse(status_code=sendResponse.status_code, content="Send Response Error")
        return JSONResponse(status_code=200, content={"message": GPTresponse})
    except:
        return JSONResponse(status_code=500, content={"error": "Internal Server Error"})

File: Database/src/routers/message.ts
import express from 'express'
import { createSchema } from '../schemas/message'
import Chat from '../models/chat'
import Message from '../models/message'

const router = express.Router()

router.post("/create", async (req, res) => {
    try {
        const createData = createSchema.safeParse(req.body)
        if (!createData.success) {
            return res.status(400).send('Body not match')
        }
        const chat = await Chat.findById(createData.data.chatid);
        if (chat) {
            const msg = await Message.create({
                username: createData.data.username,
                content: createData.data.content,
                isNotGPT: createData.data.isNotGPT,
                type: createData.data.type,
                chatid: createData.data.chatid,
                createdAt: new Date(),
                updatedAt: new Date(),
            })
            chat.messages.push(msg._id);
            const response = await chat.save();
            return res.status(200).send(response);
        } else {
            return res.status(400).send('Chat not found');
        }
    }
    catch (error) {
        return res.status(500).send(error);
    }
})

export default router
